HPCC-25477 Add ECLWatch UI testing into Smoketest GitHub action

AttilaVamos opened the pull request:
Implement Selenium based ECLWatch UI tests
    
Dependencies:
    - full platform build, install and start-up
    - install Selenum and Chrome client
    
Precompile provided Java testcodes from esp/src/test-ui/tests/ directory

Execute them with provided run.sh from es/src/test-ui/ directory
    
Capture stdout and stderr of run.sh and create artifact if any test fails.
  
Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->
    Tested locally.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-14722, label: HPCC-25477-impr-7.12.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 87967cedd56a4060397b24a4b4907dee09e570eb
Estimated completion time is ~0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-82.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-82.ca-central-1.compute.internal
Sha: 87967cedd56a4060397b24a4b4907dee09e570eb
Build: success
Milestone:Install hpccsystems-platform-community_7.12.39-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  591 sec (00:09:51) |  107 sec (00:01:47) |   20 sec (00:00:20) |   16 sec (00:00:16) |    0 sec (00:00:00) |   15 sec (00:00:15) |  758 sec (00:12:38) |

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+
+  # NB: this doesn't really need the post-setup data files included in install.tgz
+  # but as this is relatively quick and in parallel with others, it probably doesn't matter
+  eclwatchuitests:
```
trivial: for consistency/readability can you make this eclwatch-ui-tests

AttilaVamos replied:
Changed

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+          libhiredis-dev libtbb-dev libxalan-c-dev libnuma-dev nodejs libevent-dev libatlas-base-dev libblas-dev python3-dev \
+          default-libmysqlclient-dev libsqlite3-dev libmemcached-dev libcurl4-openssl-dev pkg-config libtool autotools-dev automake \
+          libssl-dev gdb
+        wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
```
minor: no harm adding a new step (e.g. name: ui-prerequisites), and would make it slightly clearer that the existing Prerequisites is same as in other sub-jobs.

AttilaVamos replied:
New step added

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+        export CLASSPATH=".:../selenium-server-standalone-3.141.59.jar:../testng-6.8.7.jar"
+        if [[ -d install/test-ui/tests ]]
+        then
+                javaFilesCount=$(find install/test-ui/tests/ -iname '*.java' -type f -print | wc -l )
```
indentation different/inconsistent with elsewhere in this action, please indent by 2 spaces

AttilaVamos replied:
Indentation is fixed

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+                                echo "No run.sh, so do nothing."
+                        fi
+                else
+                        echo "No java files, do noting."
```
typo: noting->nothing

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+        wget https://selenium-release.storage.googleapis.com/3.141/selenium-server-standalone-3.141.59.jar
+        wget http://www.java2s.com/Code/JarDownload/testng/testng-6.8.7.jar.zip
+        unzip testng-6.8.7.jar.zip
+        export CLASSPATH=".:../selenium-server-standalone-3.141.59.jar:../testng-6.8.7.jar"
```
I don't think this CLASSPATH line needed here, as you redeclare in next step. before actually needed

AttilaVamos replied:
Agree, removed

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+                        cp -v install/test-ui/tests/*java tests/
+                        cd tests
+                        javac *.java
+                        cd ..
```
Why are they copied out of install/test-ui/tests/ into ./tests/ before running, rather than just run them where they were?

AttilaVamos replied:
This and all run.sh related issue comes from the original container based solution. 
Copy removed.

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+        source install/opt/HPCCSystems/sbin/hpcc_setenv
+        install/opt/HPCCSystems/etc/init.d/hpcc-init start
+        mkdir tests
+        export CLASSPATH=".:../selenium-server-standalone-3.141.59.jar:../testng-6.8.7.jar"
```
those relative paths are a little confusing, I can see end up being right because you cd into tests, but..
would be clearer if used absolute, e.g. used $(realpath selenium-server-standalone-3.141.59.jar)

AttilaVamos replied:
Changed

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+                        # get local IP
+                        LOCAL_IP4=$(ip -o -4 addr list eth0 | awk '{print $4}' | cut -d/ -f1)
+                        export LOCAL_IP4
+                        echo "LOCAL_IP4: $LOCAL_IP4"
```
do really need this, or could you just "localhost" in the URL?

AttilaVamos replied:
Checked with localhost. It is working, so changed.

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+                                cp -v install/test-ui/run.sh .
+                                ./run.sh http://$LOCAL_IP4:8010 > eclWatchUiTest.log 2>&1
+                                echo "UI test done"
+                                exit $( egrep -c '^Fail$' ./eclWatchUiTest.log )
```
Is there any guarantee that "^Fail$" is part of the result?
It looks like run.sh is designed to return 0 on success and anything else should be used as the indicator of failure.
So think this should be changed to return error code from run.sh

AttilaVamos replied:
Changed

jakesmith commented on esp/src/test-ui/run.sh:
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+cd ./tests
```
I think this should be removed, not assume in script to be where tests are,
and instead the current dir used.
This is probably related to the copying of the tests to ./tests, which could be avoided.

AttilaVamos replied:
Removed and reworked.

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,84 @@ jobs:
+                        LOCAL_IP4=$(ip -o -4 addr list eth0 | awk '{print $4}' | cut -d/ -f1)
+                        export LOCAL_IP4
+                        echo "LOCAL_IP4: $LOCAL_IP4"
+                        if [[ -f install/test-ui/run.sh ]]
```
given run.sh is being added in this PR itself, this doesn't feel necessary.
(we don't check every file we use to see if missing)

AttilaVamos replied:
Changed

jakesmith commented:
@AttilaVamos - looks fine overall, but I think could be simplified.
I think more of the boiler-plate code for running the test should move into the script (run.sh or similar), so the github action is less verbose.
i.e. smoketest just calls the UI test script after setup (starting platform, settings classPath), and either gives run.sh the path where tests are (in install/..) or uses current directory after smoketest has cd there.

HPCCSmoketest commented:
Process of PR-14722, label: HPCC-25477-impr-7.12.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: ea1a0e8a7d06928036e661f349c789c92268825d
Estimated completion time is ~0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-21.ca-central-1.compute.internal

jakesmith commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,86 @@ jobs:
+    - name: Check-ECLWatch-UI-Test-Directory
+      run: |
+        tar --zstd -xf install.tgz
+        if [[ ! -d install/test-ui ]]
```
trivial: woud have been enough just to test [[ ! -d install/test-ui/tests ]] , simplifying/reducing script.

AttilaVamos replied:
Ok

jakesmith reviewed (commented):
@AttilaVamos looks good, a couple of minor points.

Could be merged as is, but it will run all steps unconditionally at the moment, which means that if the checks for dirs being missing were hit, it would fail in 'run' step.

AttilaVamos commented on .github/workflows/smoketest.yml:
```diff
@@ -449,3 +450,86 @@ jobs:
+        if [[ ! -d install/test-ui ]]
+        then
+          echo "ECLWatch test-ui directory missing"
+          return 0
```
Ok, I see. I fix it.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-21.ca-central-1.compute.internal
Sha: ea1a0e8a7d06928036e661f349c789c92268825d
Build: success
Milestone:Install hpccsystems-platform-community_7.12.41-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  593 sec (00:09:53) |  104 sec (00:01:44) |   21 sec (00:00:21) |   18 sec (00:00:18) |    0 sec (00:00:00) |   15 sec (00:00:15) |  759 sec (00:12:39) |

HPCCSmoketest commented:
Process of PR-14722, label: HPCC-25477-impr-7.12.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 39ee75dc1632af11922e6663636f1975c5e208ec
Estimated completion time is ~0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-77.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-77.ca-central-1.compute.internal
Sha: 39ee75dc1632af11922e6663636f1975c5e208ec
Build: success
Milestone:Install hpccsystems-platform-community_7.12.41-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  602 sec (00:10:02) |  107 sec (00:01:47) |   20 sec (00:00:20) |   16 sec (00:00:16) |    0 sec (00:00:00) |   15 sec (00:00:15) |  769 sec (00:12:49) |

HPCCSmoketest commented:
Process of PR-14722, label: HPCC-25477-impr-7.12.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: 5cd1f7e0bb35f30a92142f95fe5d0f9b2633ee8a
Estimated completion time is ~0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-22.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-22.ca-central-1.compute.internal
Sha: 5cd1f7e0bb35f30a92142f95fe5d0f9b2633ee8a
Build: success
Milestone:Install hpccsystems-platform-community_7.12.41-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  593 sec (00:09:53) |  105 sec (00:01:45) |   20 sec (00:00:20) |   16 sec (00:00:16) |    0 sec (00:00:00) |   15 sec (00:00:15) |  757 sec (00:12:37) |

jakesmith reviewed (commented):
@AttilaVamos looks good, please squash

AttilaVamos committed:
HPCC-25477 Add ECLWatch UI testing into Smoketest GitHub action

Implement Selenium based ECLWatch UI tests

Dependencies:
- full platform build, install and start-up
- install Selenum and Chrome client

Precompile provided Java testcodes from esp/src/test-ui/tests/ directory

Execute them with provided run.sh from es/src/test-ui/ directory

Capture stdout and stderr of run.sh and create artifact if any test fails.

Tested locally.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

AttilaVamos commented:
@jakesmith Squashed.

HPCCSmoketest commented:
Process of PR-14722, label: HPCC-25477-impr-7.12.x starts now.
The reason of this test is: Forced to re-test.
Commit ID: c6d211335273502a0223adb5d8c28c6d61d7f09e
Estimated completion time is ~0.15 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-86.ca-central-1.compute.internal

jakesmith reviewed (approved):
@AttilaVamos - looks good

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-86.ca-central-1.compute.internal
Sha: c6d211335273502a0223adb5d8c28c6d61d7f09e
Build: success
Milestone:Install hpccsystems-platform-community_7.12.41-closedown0.el7.x86_64.rpm
HPCC Start: OK
HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  595 sec (00:09:55) |  107 sec (00:01:47) |   20 sec (00:00:20) |   16 sec (00:00:16) |    0 sec (00:00:00) |   15 sec (00:00:15) |  761 sec (00:12:41) |

AttilaVamos commented:
@richardkchapman please merge.
